sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuAvroScan.scala

```diff
@@ -0,0 +1,472 @@
+case class GpuAvroPartitionReaderFactory(
+    sqlConf: SQLConf,
+    broadcastedConf: Broadcast[SerializableConfiguration],
+    dataSchema: StructType,
+    readDataSchema: StructType,
+    partitionSchema: StructType,
+    @transient rapidsConf: RapidsConf,
+    ignoreExtension: Boolean,
+    metrics: Map[String, GpuMetric]) extends FilePartitionReaderFactory with Logging {
+
+  private val debugDumpPrefix = rapidsConf.parquetDebugDumpPrefix
+  private val maxReadBatchSizeRows = rapidsConf.maxReadBatchSizeRows
+  private val maxReadBatchSizeBytes = rapidsConf.maxReadBatchSizeBytes
+
+  override def buildReader(partitionedFile: PartitionedFile): PartitionReader[InternalRow] = {
+    throw new IllegalStateException("ROW BASED PARSING IS NOT SUPPORTED ON THE GPU...")
+  }
+
+  override def buildColumnarReader(partFile: PartitionedFile): PartitionReader[ColumnarBatch] = {
+    val conf = broadcastedConf.value.value
+    val blockMeta = GpuAvroFileFilterHandler(sqlConf, broadcastedConf,
+      ignoreExtension, broadcastedConf.value.value).filterBlocks(partFile)
+    val reader = new PartitionReaderWithBytesRead(new AvroPartitionReader(conf, partFile, blockMeta,
+      readDataSchema, debugDumpPrefix, maxReadBatchSizeRows,
+      maxReadBatchSizeBytes, metrics))
+    ColumnarPartitionReaderWithPartitionValues.newReader(partFile, reader, partitionSchema)
+  }
+}
+
+/**
+ * A tool to filter Avro blocks
+ *
+ * @param sqlConf         SQLConf
+ * @param broadcastedConf the Hadoop configuration
+ */
+private case class GpuAvroFileFilterHandler(
+    @transient sqlConf: SQLConf,
+    broadcastedConf: Broadcast[SerializableConfiguration],
+    ignoreExtension: Boolean,
+    hadoopConf: Configuration) extends Arm with Logging {
+
+  def filterBlocks(partFile: PartitionedFile): AvroBlockMeta = {
+
+    def passSync(blockStart: Long, position: Long): Boolean = {
+      blockStart >= position + SYNC_SIZE
+    }
+
+    if (ignoreExtension || partFile.filePath.endsWith(".avro")) {
+      val in = new FsInput(new Path(new URI(partFile.filePath)), hadoopConf)
+      closeOnExcept(in) { _ =>
+        withResource(AvroDataFileReader.openReader(in)) { reader =>
+          val blocks = reader.getBlocks()
+          val filteredBlocks = new ArrayBuffer[BlockInfo]()
+          blocks.foreach(block => {
+            if (partFile.start <= block.blockStart - SYNC_SIZE &&
+              !passSync(block.blockStart, partFile.start + partFile.length)) {
+              filteredBlocks.append(block)
+            }
+          })
+          AvroBlockMeta(reader.getHeader(), filteredBlocks)
+        }
+      }
+    } else {
+      AvroBlockMeta(new Header(), Seq.empty)
+    }
+  }
+}
+
+/**
+ * Avro block meta info
+ *
+ * @param header the header of avro file
+ * @param blocks the total block info of avro file
+ */
+case class AvroBlockMeta(header: Header, blocks: Seq[BlockInfo])
+
+/**
+ * CopyRange to indicate from where to copy.
+ *
+ * @param offset
+ * @param length
```

Comment:
Add descriptions for these or remove the comments for them.

Reply:
Done